lib/compress/zstd_compress.c

```diff
@@ -2903,6 +2977,24 @@ static size_t ZSTD_buildSeqStore(ZSTD_CCtx* zc, const void* src, size_t srcSize)
         }
         if (zc->externSeqStore.pos < zc->externSeqStore.size) {
             assert(zc->appliedParams.ldmParams.enableLdm == ZSTD_ps_disable);
+
```

Comment:
I asked about this offline a while back.

@terrelln:
> It is used by multi-threaded LDM
> We generate the LDM sequences externally because it is a serial process, then inject them into the cctx so the normal block finding can happen in parallel.

The discussion a while back also indicated that the block splitter can trigger this branch.

I'm personally not familiar with why this branch exists, so I disabled the external matchfinder API across the board here. If the block-splitter indeed triggers this branch, then the external matchfinder API will fail compression at high compression levels.

We could add some logic during cParam validation to disable block splitting if the external matchfinder is enabled or vice versa. How would you like to proceed?

Reply:
Yeah, in current situation, setting an external matchfinder with a high compression level is an automatic compression failure.
This might be acceptable on the short term, but we certainly don't want that to remain "long term", aka it can't remain the situation at release time.

When it comes to having a cleaner situation for release, multiple solutions are possible, and I don't have strong opinion on any of them. 
It starts from making external matchfinder mode compatible with ldm and block splitter, though it certainly requires some work.
A shorter work around could be, as you describe, to automatically change these parameters to "disable" when external matchfinder is set. 
If we go this route, another potential option could be to _ignore_ these parameters when an external matchfinder is set.

Any of these solutions is probably better than an automatic failure when combining high compression levels with external match finder. Each one comes with its own set of trade offs, so there is room for discussion.

Reply:
I tested, and it turns out this branch is not related to the block splitter: https://gist.github.com/embg/56d42da63eb2eeb86259eeefb0d6b2a4

Therefore, the existing code will not fail at high compression levels due the the block splitter being enabled. Compression will only fail if LDM is enabled (either single-thread or multi-thread LDM).

Reply:
Good news !